### ISSUE TITLE
Incidents to Claims TimeZone

### ISSUE THREAD

**Issue by mboneil10** [umts/incidents#180]
An incident's `incident_date`  in the sql insert statement to claims is expected to be in EDT, but it's in GMT. It was requested that the `incident_date` stays in EDT.

From an email I'd sent to Liza:
> The following is the an insert statement on one incident. I made an incident in development that has the following fields to be exported to claims:
> ```
> {:incident=>
>   {"DateEntered"=>"2018-11-26",
>    "IncidentDate"=>Sat, 27 Oct 2018 19:03:00 EDT -04:00,
>    :street=>"Hackett Heights",
>    :city=>"Northampton",
>    :state=>"MA",
>    :zip=>"01934",
>    :longitude=>"-72.22564",
>    :latitude=>"42.21163",
>    "Company"=>2,
>    "IncidentDesc"=>nil,
>    "EmployeeID"=>"2781",
>    "Driver"=>nil,
>    "DriverDesc"=>
>     "Consequuntur tempore qui iste omnis. Optio qui qui qui odio. Reprehenderit sapiente totam sunt corrupti.\nVelit nisi et consequuntur facere sed sequi et. Voluptas aliquam nisi quam vitae. Eos voluptatem nihil pariatur dolores sit vel eum. Distinctio corporis a doloribus ullam voluptatibus.\nArchitecto officia veniam quia ut delectus est odit. Dignissimos ut alias rem magnam ea. Sint ut est asperiores aliquam consequatur quibusdam rerum voluptatem. Modi debitis et quod voluptates sunt. Enim totam maiores similique aut est.\nIpsum consequatur sint aut doloribus qui. Qui dolor dicta dolorem corrupti recusandae molestiae. Voluptas eos provident facilis explicabo sed consequatur sunt vero. Quo distinctio dolorem culpa explicabo cupiditate.\nEos sint ex quod at dolorum. Labore tempore iste sed aut occaecati. Numquam delectus consequatur unde voluptate. Itaque pariatur ut adipisci officia sunt.",
>    "VehicleRouteNum"=>"637",
>    "VehicleNum"=>nil,
>    "VehicleDamageArea"=>"Scratches",
>    "PointOfContact"=>"Scratches",
>    "Status"=>"ir",
>    :reason1=>"C-1",
>    :reason2=>"D-51"},
>  :drivers_report=>
>   {"FileID"=>nil,
>    "Citation"=>false,
>    "Weather"=>"Raining",
>    "SurrCond"=>"Slushy",
>    "Lighting"=>"Dawn/Dusk",
>    "Speed"=>36,
>    "PointOfContact"=>"Scratches",
>    "TotalPass"=>6,
>    "Ambulance"=>false,
>    "OVTow"=>false,
>    "PVTATow"=>false}}
> ```
> 
> The generated SQL statement (incident in claims) looks like this:
> 
> ```
> SQL (1.1ms)  INSERT INTO `incident` (`DateEntered`, `IncidentDate`, `street`, `city`, `state`, `zip`, `longitude`, `latitude`, `Company`, `EmployeeID`, 
> `DriverDesc`, `VehicleRouteNum`, `VehicleDamageArea`, `PointOfContact`, `Status`, `reason1`, `reason2`) VALUES ('2018-11-26', '2018-10-27 23:03:00', 
> 'Hackett Heights', 'Northampton', 'MA', '01934', '-72.22564', '42.21163', 2, 2781, 'Consequuntur tempore qui iste omnis. Optio qui qui qui odio. Reprehenderit 
> sapiente totam sunt corrupti.\nVelit nisi et consequuntur facere sed sequi et. Voluptas aliquam nisi quam vitae. Eos voluptatem nihil pariatur dolores sit vel eum. 
> Distinctio corporis a doloribus ullam voluptatibus.\nArchitecto officia veniam quia ut delectus est odit. Dignissimos ut alias rem magnam ea. Sint ut est asperiores 
> aliquam consequatur quibusdam rerum voluptatem. Modi debitis et quod voluptates sunt. Enim totam maiores similique aut est.\nIpsum consequatur sint aut 
> doloribus qui. Qui dolor dicta dolorem corrupti recusandae molestiae. Voluptas eos provident facilis explicabo sed consequatur sunt vero. Quo distinctio dolorem 
> culpa explicabo cupiditate.\nEos sint ex quod at dolorum. Labore tempore iste sed aut occaecati. Numquam delectus consequatur unde voluptate. Itaque pariatur 
> ut adipisci officia sunt.', '637', 'Scratches', 'Scratches', 'ir', 'C-1', 'D-51')
> ```
> 
> The incident date in claims is now 23:03:00 instead of 19:03:00 EDT -04:00.

**Comment by dfaulken** [comment]
It's sending in UTC. It's not "instead of", it's just local to a different time zone. Maybe there's a way to ask ActiveRecord to include full ISO8601-spec datetimes? Or to parse it as UTC on the claims end?

**Comment by mboneil10** [comment]
Ah, ok. I don't quite understand it. I'll send you the most recent email thread between me and Liza.

**Comment by dfaulken** [comment]
My two suggestions here are still my two suggestions. If this is still an issue, which I have no reason to believe that it isn't, we should be able to rectify either by:

1. asking Liza to make a code change to the claims database to parse the times as being in EST/EDT (less good); or
2. finding some way for the method to include full ISO-8601-spec datetimes, at which point hopefully the claims software would be able to correctly interpret the time offset, and therefore the correct local time, of each incident.

In either case, we need to coordinate the change with Liza.

**Comment by dfaulken** [comment]
In a fun twist of events, it seems that we never communicated to any users that the claims integration was actually active, and neither Jamin nor Dave were aware that the integration was live. The last communication that I could find was from me, in April 2018, saying that the button was there but that it would not affect production claims.

Twisting further, we tested the integration and found that it is currently inoperative. Rails cannot initiate a connection to the claims production database. Liza was the last person to use it when she verified functionality some months ago.

On the bright side, when Jamin tested the claims integration, it did produce an XML file, as we were hoping to test in parallel. @sherson and I agreed that the best path forward is to work with Lim and Jiza to put the XML into production, rather than working to restore the direct database connection, particularly given that no one is currently relying on the integration today anyway.

I'll reach out to Jim and Liza.

**Comment by dfaulken** [comment]
Emailed Liza and Jim today.

**Comment by dfaulken** [comment]
Jim on 6/11:

David

I met with Liza this morning to discuss this process. 

Would it be possible to drop the XML files to a directory and make them accessible via HTTP? 

Also would it be possible to add an increment number to each filename? Example incident_1.xml, incident_2.xml  

Is it necessary to alert incidents if these imports fail/succeed or is it OK to keep that on the Claims side? 
 
If yes than Liza would just use code on the Claims side to check for new files every 10 minutes. I would then setup a cron job to archive the XML files to a different directory on a scheduled basis.

Thanks for all your work on this.
Jim

**Comment by dfaulken** [comment]
Me on 6/11:

Hi Jim,

Thanks for looking into this.

Yes, having a unique number for each incident is no problem. Unless you mean that the export files themselves should have an incrementing counter? This is more complicated since not all incidents are exported to claims, so we'd need to have some other counter somewhere. I'll think about how to do that. Let me know if that's a hard requirement.

No, not strictly necessary for incidents to be aware of whether a claims import has succeeded. Anyone who is exporting to claims would be moving right over to claims to finish up any data entry anyway, so it would be pretty evident pretty quickly if an import failed.

I'll have to look into what making exports available via HTTP would look like. This introduces some additional complication that we wouldn't have with an SFTP transfer.

For instance, it may be simpler, instead of having the web application framework handle it, to just serve up that particular directory with Apache, but then the complicated part would be the routing.

We also need to think about authentication, since there's sensitive data involved, and not anyone who finds the URL should be able to access these resources.

I'll look into what seems simplest and I'll let you know what I come up with.

Dave

**Comment by dfaulken** [comment]
The logical next step here would be a design session between UMTS IT and PVTA IT to discuss options.

**Comment by sherson** [comment]
I think we can close this since we're giving them XML file (rather than generating SQL). If there are problems with the date format, we can reopen this (or open a new issue).
